File: trie/create_and_search_in_trie.ts
class TrieNode {
    is_completed: boolean;
    child: Map<String, TrieNode>;
    
    constructor() {
        this.child = new Map();
        this.is_completed = false;

    }
};

// create root node
const root = new TrieNode();



function add() {
    const strings: string[] = ["abc", "abgl", "cdf", "abcd", "lmn"];

    for (let str of strings) {

        console.log('str', str);
        
        
        let currNode = root;
        
        for (let ch of str) {
            console.log('ch', ch);
            
            let nextNodeAdd = currNode.child.get(ch);
    
            console.log('\n--- nextNodeAdd', nextNodeAdd);
    
            if (nextNodeAdd === undefined) { // character is not present
    
                let newNode = new TrieNode();
    
                currNode.child.set(ch, newNode);
                currNode = newNode;
            } else {
                currNode = nextNodeAdd;
            }
    
            console.log('\n--- currNode', currNode);
    
            currNode.is_completed = true;
        }
    }
}

function search() {
    const strings: string[] = ["abc", "abgl", "cdf", "abcd", "lmn", "md"];

    const matchFound: boolean[] = [];

    for (let str of strings) {
        let currNode = root;

        let isFound = true;

        for (let ch of str) {
            console.log('ch', ch);

            let nextNodeAdd = currNode.child.get(ch);

            if (nextNodeAdd === undefined) {
                isFound = false;        
                break;
            }

            currNode = nextNodeAdd;
        }

        if (currNode.is_completed === false || isFound === false) {
            matchFound.push(false);
        } else {
            matchFound.push(true);
        }
    }

    for (let i=0; i<strings.length; i++) {
        console.log(strings[i], matchFound[i]);
    }
}

function isStringWithPrefix(prefix: String) {
    let currNode = root;

    let isFound = true;

    for (let ch of prefix) {
        console.log('ch', ch);

        let nextNodeAdd = currNode.child.get(ch);

        if (nextNodeAdd === undefined) {
            isFound = false;
            break;
        }

        currNode = nextNodeAdd;
    }

    console.log('isFound', isFound);
}

add();
search();
isStringWithPrefix("ab");

File: trie/create_and_search_in_trie.cpp
# include <bits/stdc++.h>

using namespace std;

class TrieNode {
    public:
        unordered_map<char, TrieNode*> children;
        bool is_completed;

    TrieNode() {
        this->children = unordered_map<char, TrieNode*>{};
        this->is_completed = false;
    }
};

TrieNode *root = new TrieNode();

void add() {
    vector<string> strings = {"abc", "abgl", "cdf", "abcd", "lmn"};

    for (auto s: strings) {
        TrieNode *currNode = root;

        for (int i = 0; i < s.length(); i++) {
            char ch = s[i];

            if (currNode->children.count(ch) == 0) { // currNode->children.find(ch) == currNode->children.end()

                TrieNode* nextNode = new TrieNode();

                // (currNode->children).insert(make_pair(ch, nextNode));
                currNode->children[ch] = nextNode;
                currNode = nextNode;
            } else {
                currNode = currNode->children[ch];
            }

            currNode->is_completed = true;
        }
    }
}

void search() {
    vector<string> strings = {"abc", "abgl", "cdf", "abcd", "lmn", "lmno", "plw"};

    vector<bool> matches;

    for (auto s: strings) {
        TrieNode *currNode = root;

        bool isFound = true;

        for (int i = 0; i < s.length(); i++) {
            char ch = s[i];

            if (currNode->children.count(ch) == 0) { // currNode->children.find(ch) == currNode->children.end()
                isFound = false;
                break;
            } else {
                currNode = currNode->children[ch];
            }
        }

        if (currNode->is_completed == false || isFound == false) {
                matches.push_back(false);
        } else {
            matches.push_back(true);
        }
    }

    for (int i = 0; i < strings.size(); i++) {
        cout << strings[i] << matches[i] << endl;
    }
}

int main() {
    add();
    search();

    return 0;
}